### PR TITLE
WIP: Shwap prototype

### PR DESCRIPTION
**Pull request by walldiss**
Prototype for shwap protocol and storage. Contain all changes required for shwap implementation. This PR is not meant for merge. All changes will be extracted from prototype into smaller and nicer PRs. 

Based on: https://github.com/celestiaorg/celestia-node/pull/2675
Tracking issue: https://github.com/celestiaorg/celestia-node/issues/2971

### CODE REVIEW

File: share/shwap/data_id.go [hunk 1]
```diff
@@ -17,7 +17,6 @@ const DataIDSize = RowIDSize + share.NamespaceSize
 
 // DataID is an unique identifier of a namespaced Data inside EDS Row.
 type DataID struct {
-	// TODO(@walldiss): why embed instead of just having a field?
```
Comment (Wondertan) [inline share/shwap/data_id.go]: so you don't have to type .RowID to access its fields and allow passing DataID so where its RowID is expected. Also, we don't need to define GetHeight for every type and doing that on the new EdsID is enough

File: share/shwap/data_hasher_test.go [hunk 2]
```diff
@@ -10,8 +10,6 @@ import (
 	"github.com/celestiaorg/celestia-node/share/sharetest"
 )
 
-// TODO(@walldiss):
-// FIX: hasher test succeed, while logging unmarshal error: "unmarshaling Data: proto: Data: wiretype end group for non-group"
```
Comment (Wondertan) [inline share/shwap/data_hasher_test.go]: That's by design. It tests the negative case as well

File: share/shwap/row_id.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	"github.com/ipfs/go-cid"
+	mh "github.com/multiformats/go-multihash"
+
+	"github.com/celestiaorg/rsmt2d"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/store/file"
+)
+
+// TODO(@walldiss): maybe move into separate subpkg?
```
Comment (Wondertan) [inline share/shwap/row_id.go:18]: Why?

File: share/shwap/row_id.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+// MarshalTo encodes RowID into given byte slice.
+// NOTE: Proto is avoided because
+// * Its size is not deterministic which is required for IPLD.
+// * No support for uint16
+func (rid RowID) MarshalTo(data []byte) (int, error) {
+	// TODO:(@walldiss): this works, only if data underlying array was preallocated with
+	//  enough size. Otherwise Caller might not see the changes.
+	data = binary.BigEndian.AppendUint64(data, rid.Height)
```
Comment (Wondertan) [inline share/shwap/row_id.go:84]: True

File: share/shwap/getter/getter.go [hunk 5]
```diff
@@ -1,24 +1,22 @@
...
-	"github.com/celestiaorg/celestia-app/pkg/wrapper"
-	"github.com/celestiaorg/rsmt2d"
-
 	"github.com/celestiaorg/celestia-node/header"
 	"github.com/celestiaorg/celestia-node/share"
 )
 
-// TODO: GetRow method
 type Getter struct {
+	// TODO(@walldiss): why not blockservice?
```
Comment (Wondertan) [inline share/shwap/getter/getter.go:19]: Because we don't need to cache on disk in most cases

File: share/shwap/getter/getter.go [hunk 6]
```diff
@@ -218,56 +197,62 @@ func (g *Getter) GetSharesByNamespace(
...
-		nShrs = append(nShrs, share.NamespacedRow{
-			Shares: row.DataShares,
-			Proof:  &proof,
-		})
+	// only persist when all samples received
+	if len(blks) != len(cids) {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		return nil, fmt.Errorf("not all shares were found")
```
Comment (Wondertan) [inline share/shwap/getter/getter.go:254]: blocks, because now its generalized

File: share/shwap/getter/getter.go [hunk 7]
```diff
@@ -218,56 +197,62 @@ func (g *Getter) GetSharesByNamespace(
...
-			return err
+			return nil, fmt.Errorf("getting row from block: %w", err)
 		}
 
-		nsStartIdx := dids[0].RowIndex
-		idx := d.RowIndex - nsStartIdx
-		datas[idx] = d
-		return nil
+		if data.RowIndex < uint16(from) || data.RowIndex >= uint16(to) {
+			// should never happen, because rows should be verified against root by the time they are returned
```
Comment (Wondertan) [inline share/shwap/getter/getter.go:223]: Why do we need if then, if its never happens?

File: share/shwap/getter/getter.go [hunk 8]
```diff
@@ -138,58 +121,54 @@ func (g *Getter) GetEDS(ctx context.Context, hdr *header.ExtendedHeader) (*rsmt2
...
-	verifyFn := func(row Row) error {
-		err := row.Verify(hdr.DAH)
-		if err != nil {
-			return err
+	if len(blks) != sqrLn/2 {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
 		}
+		return nil, fmt.Errorf("not all rows were found")
+	}
```
Comment (Wondertan) [inline share/shwap/getter/getter.go]: duplicates logic in getBlocks so can be removed

File: share/shwap/sample_id.go [hunk 9]
```diff
@@ -127,3 +133,7 @@ func (sid SampleID) BlockFromFile(ctx context.Context, f file.EdsFile) (blocks.B
 	}
 	return blk, nil
 }
+
+func (sid SampleID) Release() {
+	sampleVerifiers.Delete(sid)
+}
```
Comment (Wondertan) [inline share/shwap/sample_id.go:136]: little reminder to godoc this

File: nodebuilder/share/config.go [hunk 10]
```diff
@@ -18,6 +18,7 @@ type Config struct {
 	EDSStoreParams *store.Parameters
 
 	UseShareExchange bool
+	UseShareSwap     bool
```
Comment (Wondertan) [inline nodebuilder/share/config.go:21]: Better keep it Shwap and rename the above one ot shrex, so there is less confusions around namings

File: share/shwap/row_id.go (hunk 4 above)
Comment (walldiss) [reply share/shwap/row_id.go:81]: Needs to be fixed tho

**Comment by Wondertan** [comment]
We merged #3675, so can close this one

**Pull request closed without merging**